File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.py
class Node:
    def __init__(self):
        self.isEndWord = False
        self.children = defaultdict(Node)

class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        root = self.root
        for char in word:
            root = root.children[char]
        root.isEndWord = True

    def search(self, word: str) -> bool:
        root = self.root
        for char in word:
            if char not in root.children:
                return False
            root = root.children[char]
        return root.isEndWord

    def startsWith(self, prefix: str) -> bool:
        root = self.root
        for char in prefix:
            if char not in root.children:
                return False
            root = root.children[char]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
struct TrieNode {
    bool is_word;
    TrieNode *child[26];
    
    TrieNode() {
        is_word = false;
        for (int i = 0; i < 26; ++i) {
            child[i] = nullptr;
        }
    }
};

class Trie {
private:
    TrieNode *root;
    
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode *curr = root;
        
        for (char c : word) {
            if (curr->child[c - 'a'] == nullptr) {
                curr->child[c - 'a'] = new TrieNode();
            }
            curr = curr->child[c - 'a'];
        }
        
        curr->is_word = true;
    }
    
    bool search(string word) {
        TrieNode *curr = root;
        
        for (char c : word) {
            if (curr->child[c - 'a'] == nullptr) {
                return false;
            }
            curr = curr->child[c - 'a'];
        }
        
        return curr->is_word;
    }
    
    bool startsWith(string word) {
        TrieNode *curr = root;
        
        for (char c : word) {
            if (curr->child[c - 'a'] == nullptr) {
                return false;
            }
            curr = curr->child[c - 'a'];
        }
        
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */